### PR TITLE
Support the Chapel bytes type in single locale Python interop

### LINKED ISSUE #14110: Ability to export routines that accept / return `byte`s
As a Chapel programmer working with Python, I'd like to be able to pass Python bytes objects into exported Chapel routines, operate on them within Chapel as type `bytes`, and return `bytes` expressions from exported Chapel routines back to Python as `bytes` objects.

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR adds support for the `bytes` type to Python interop in a manner similar to how `string` is supported today. 

The key differences between interop support for `bytes` and `string` are:

- The `bytes` type can comfortably handle embedded null characters, while the _current_ interop machinery for `string` cannot.
- The `bytes` type can hold arbitrary binary data, while the the `string` type is limited to valid UTF-8 codepoints.

Resolves #14110.

---

I'm very intentionally only adding support for Python interop with this PR, as discussion with @bradcray led me to think that perhaps a different approach might be best for the C language.

Of course, this depends on whether we want to guarantee that the C bridge code produced by `--library-python` and the C code produced by `--library` both remain the same (I would argue that they should not be required to, but that could best be discussed in another issue).

Though there is a test for the `chpl_bytes` wrapper in `test/interop/C`, that was mainly used for my own benefit and could be removed.

---

Testing (assumes `CHPL_LIB_PIC=pic`):

- [x] `ALL` on `linux64` when `CHPL_COMM=none` (no `cygwin`/`numpy`)
- [x] `ALL` on `linux64` when `CHPL_COMM=gasnet` (no `cygwin`/`numpy`)
- [x] `interop` on `darwin` when `CHPL_COMM=none`
- [x] `interop` on `darwin` when `CHPL_COMM=gasnet`
- [x] `interop` on `linux64` when `CHPL_COMM=none`
- [x] `interop` on `linux64` when `CHPL_COMM=gasnet`, `LAUNCHER=none`

### CODE REVIEW

**Comment by bradcray** [comment]
It seems I was likely mistaken if I gave the impression that Chapel strings can't handle NULL bytes.  It's been asserted to me that they can and that it's just the interop code paths that can't at present (since they use `strlen()` rather than the length stored in Chapel's string record).  I convinced myself that Chapel strings / UTF-8 strings couldn't store NULL bytes, yet without checking.

**Comment by dlongnecke-cray** [comment]
> It seems I was likely mistaken if I gave the impression that Chapel strings can't handle NULL bytes. It's been asserted to me that they can and that it's just the interop code paths that can't at present (since they use `strlen()` rather than the length stored in Chapel's string record). I convinced myself that Chapel strings / UTF-8 strings couldn't store NULL bytes, yet without checking.

I edited my OP to clarify those points, thanks.

**Comment by lydia-duncan** [comment]
> Testing (assumes CHPL_LIB_PIC=pic):
> 
> ALL on linux64 when CHPL_COMM=none
> ALL on linux64 when CHPL_COMM=gasnet
> ALL on linux64 when CHPL_COMM=none, CHPL_LLVM=llvm, --llvm

I think this is good to ensure, but also remember that the machine where we run full linux testing doesn't have Cython installed (so won't trigger the python tests even with CHPL_LIB_PIC=pic), meaning you will likely need a separate run for python testing on linux

**Comment by lydia-duncan** [comment]
> Of course, this depends on whether we want to guarantee that the C bridge code produced by --library-python and the C code produced by --library both remain the same (I would argue that they should not be required to, but that could best be discussed in another issue).

I think that doing so simplifies the implementation significantly and makes it easier to debug, but that maintaining that can and should be balanced against a good user experience.  Where exactly that line lies is worth determining on a case by case basis, so I don't know that an issue for the general topic is worthwhile unless someone disagrees with that general sentiment (which seems unlikely).  I'm curious about the plan for C support but we can discuss that offline

File: modules/internal/ExportWrappers.chpl [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    use ByteBufferHelpers;
+    var data = val.data:ByteBufferHelpers.bufferType;
+    // TODO: Are these casts safe casts?
+    var length = val.size:int(64);
+    var buflen = (val.size + 1):int(64);
+    var result = createBytesWithNewBuffer(data, length=length, size=buflen);
+    return result;
+  }
+
+} // End module "ExternalString".
```
Comment (lydia-duncan) [inline modules/internal/ExportWrappers.chpl]: This comment should be updated

File: test/interop/C/exportBytes/SKIPIF [hunk 2]
```diff
@@ -0,0 +1 @@
+CHPL_COMM != none
```
Comment (lydia-duncan) [inline test/interop/C/exportBytes/SKIPIF]: Shouldn't this just be under the test/interop/C/multilocale directory, then?

Comment (lydia-duncan) [reply test/interop/C/exportBytes/SKIPIF]: Wait, ignore me, I misread that

**Comment by dlongnecke-cray** [comment]
@lydia-duncan

I'm having a bit of a design problem here.

There are type aliases in `modules/internal/ExportWrappers.chpl` that I have the compiler resolve when doing fixups. However the types they represent are needed at codegen even if no fixups have been performed.

The problem is that if `--minimal-modules` has been set the compiler doesn't have access to that module.

What's the current state of `--library*` compiles and `--minimal-modules`? Am I safe to use logic like "only resolve type aliases if not `fMinimalModules`"?

**Comment by bradcray** [comment]
> What's the current state of --library* compiles and --minimal-modules? Am I safe to use logic like "only resolve type aliases if not fMinimalModules"?

I don't think there's any assumption that they will work together and that it'd be fine to use such logic.

**Comment by lydia-duncan** [comment]
It looks like minimal modules doesn't support strings or bytes, only c_strings.  Ideally, someone could use `--library` and `--minimal-modules` together so long as they limited themselves to only situations that `--minimal-modules` supports on its own, but I certainly don't view that as a high priority.

If "only resolve type aliases if not fMinimalModules" allows `chpl --library --minimal-modules` to work for basic types, I would not consider it a failing of that combination to not support bytes or strings and would be comfortable saying that combination is supported (as long as we wrote a test that locked it in).  If it takes more work than that, I'm okay with saying you can't use that combination of flags just yet, as long as `--minimal-modules` on its own continues to work.

**Comment by dlongnecke-cray** [comment]
My plan is to add a postprocess args check that emits an error if `fLibraryCompile` and `fMinimalModules` are both true.

This stopgap seems like a better solution than wading through code trying to get things to work, especially given that getting those two flags to play well together seems like a low priority.

**Comment by bradcray** [comment]
The goal of `--minimal-modules` is to support compiler developers who want to plumb a feature through the code without having to wrestle with all of the internal, standard, package modules.  We can worry about having `--library` and `--minimal-modules` do something reasonable together the day that a developer would find utility in that combination.  Before then, I think it would be wasted effort.

**Comment by bradcray** [comment]
How much additional work would it take for multi-locale bytes Python interop?
Has anyone tried taking Lydia's Arkouda branch and seeing whether the send/recv bytes paths can now be written to use this type?

**Pull request merged**